plugins/hfp_ag_bluez5.c

```diff
@@ -35,21 +36,30 @@
 #include <ofono/plugin.h>
 #include <ofono/log.h>
 #include <ofono/modem.h>
+#include <ofono/handsfree.h>
+#include <ofono/handsfree-audio.h>
 
 #include "hfp.h"
 #include "bluez5.h"
+#include "bluetooth.h"
 
 #ifndef DBUS_TYPE_UNIX_FD
 #define DBUS_TYPE_UNIX_FD -1
 #endif
 
 #define HFP_AG_EXT_PROFILE_PATH   "/bluetooth/profile/hfp_ag"
+#define BT_ADDR_SIZE 18
 
 static guint modemwatch_id;
 static GList *modems;
 static GHashTable *sim_hash = NULL;
 static GHashTable *connection_hash;
 
+struct device_info {
+	int fd;
+	struct ofono_handsfree_card *card;
+};
+
 static void connection_destroy(gpointer data)
 {
 	int fd = GPOINTER_TO_INT(data);
```

Comment:
This function does not receive a fd anymore, but a device_info structure. And ofono_handsfree_card_remove() needs to be called when the struct is freed.